src/validation.cpp

```diff
@@ -53,11 +53,60 @@
 /**
  * Global state
  */
+namespace {
+struct CBlockIndexWorkComparator
+{
+    bool operator()(CBlockIndex *pa, CBlockIndex *pb) const {
+        // First sort by most total work, ...
+        if (pa->nChainWork > pb->nChainWork) return false;
+        if (pa->nChainWork < pb->nChainWork) return true;
+
+        // ... then by earliest time received, ...
+        if (pa->nSequenceId < pb->nSequenceId) return false;
+        if (pa->nSequenceId > pb->nSequenceId) return true;
+
+        // Use pointer address as tie breaker (should only happen with blocks
+        // loaded from disk, as those all have id 0).
+        if (pa < pb) return false;
+        if (pa > pb) return true;
+
+        // Identical blocks.
+        return false;
+    }
+};
+};
+
+/**
+ * CChainState stores and provides an API to update our local copy of the chain
+ *
+ * It generally provides access to the current chain tree, as well as functions
+ * to provide it new data, which it will appropriately validate and incorporate
+ * in its state as neccessary.
+ *
+ * Eventually, the API here is targeted at being exposed externally as a
+ * consumable libconsensus library, so any functions added must only call
+ * external functions via callbacks, or to read/write data from disk.
+ */
+class CChainState {
+public:
+    CChain chainActive;
+    BlockMap mapBlockIndex;
+    std::multimap<CBlockIndex*, CBlockIndex*> mapBlocksUnlinked;
+    std::set<CBlockIndex*, CBlockIndexWorkComparator> setBlockIndexCandidates;
+    CBlockIndex *pindexBestInvalid;
+
+    bool Load(CBlockTreeDB& blocktree);
```

Comment:
Load adds to mapBlocksUnlinked and InsertBlockIndex into mapBlockIndex. In the next PR mapBlockIndex and chainActive will be const when accessed via the global references, so would prefer to leave it.